### DISCUSSION TITLE
How to update a subscription product with a selected variant using graphQL?

### DISCUSSION THREAD

**Discussion by tarikrital** [swellstores/community#564] [Help] [answered]
Welcome,

The following code some product worked and others not, any reference for this operation?

```
mutation replaceSubscriptionProduct($id: String, $productId: ID,
  $variantId: ID,
  $qty: Int, $interval: String, $intervalCount: Int) {
  updateSubscription(
    id: $id
    input: {
      productId: $productId
      # items: { productId: $productId, quantity: $qty }
     quantity: $qty
      # options: {
      #   value:"Red"
      #   name: "Colour"
      # }
     variantId: $variantId
      billingSchedule: {
        interval: $interval
        intervalCount: $intervalCount
      }
    }
  ) {
    product {
      name
      id
      price
    }
    quantity
    trial
    grandTotal
  }
}
```

**Comment by ericingram** [comment]
What do you mean by "some product worked and others not"?

**Reply by tarikrital** [reply]
The mutation work for it before but now it doesn't! showing the error below?

```
{
  "errors": [
    {
      "message": {
        "errors": {
          "items.options": {
            "message": "Product option value is not available (Blue)",
            "code": "INVALID",
            "id": "6691b18f587ef300121585d8",
            "params": [
              {
                "id": "659f10ba65f8ee2ef3100d70",
                "value": "Blue",
                "name": "Toothbrush Colour",
                "variant": true,
                "value_id": "659f117f65f8ee2ef3100d77"
              }
            ]
          }
        },
        "path": "/subscriptions/65afe7abc174ff001254c57f"
      },
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ]
    }
  ],
  "data": {
    "updateSubscription": null
  }
}
```

**Reply by logeshswell** [reply]
Can you fix the error: Product option value is not available (Blue)?

It looks like `productOptionValue (659f117f65f8ee2ef3100d77)` is incorrect.

**Reply by logeshswell** [reply]
Can you share your graphQL mutation query for me to review?

**Reply by logeshswell** [reply]
```graphql
mutation replaceSubscriptionProduct {
  updateSubscription(
    id: "65afe7abc174ff001254c57f"
    input: {
      productId: "6691b6ecc567a40012038f7f"
      variantId: "6691b6edc567a40012038f86"
      billingSchedule: {
        interval: "monthly"
        intervalCount: 6
      }
    }
  ) {
    product {
      name
      id
      price
    }
    quantity
    trial
    grandTotal
  }
}
```
Can you try the above query and let me know if it works?

**Reply by tarikrital** [reply]
![image](https://github.com/user-attachments/assets/7b009859-2804-49de-a60d-8d9893a1305b)

Your mutation worked! why not having quantity? what is the problem?

**Reply by logeshswell** [reply]
That's not a problem. You can also pass `quantity` value in the mutation.
For more details on the query variables, please check out **Schema** page in the graphQL playground.

**Reply by tarikrital** [reply]
![image](https://github.com/user-attachments/assets/f549f524-1c90-4176-9528-0a1aa3fa01c8)

I don't see where is the problem in the query!

![image](https://github.com/user-attachments/assets/ecf656bb-13fe-4df7-891e-1f7188751fab)

The variant id is correct for `mustard` which exists but why the error showing other variant of other product I'm not using! is this a cache error!

**Reply by logeshswell** [reply]
Can you try using the same mutation with `input.quantity` field?

**Reply by tarikrital** [reply]
The same issue!

![image](https://github.com/user-attachments/assets/ff979da1-6221-4590-b073-a4c7c4f9a83c)

**Reply by logeshswell** [reply]
Sorry, I mean the below mutation:
```graphql
mutation replaceSubscriptionProduct {
  updateSubscription(
    id: "65afe7abc174ff001254c57f"
    input: {
      productId: "6691b6ecc567a40012038f7f"
      variantId: "6691b6edc567a40012038f86"
      quantity: 3
      billingSchedule: {
        interval: "monthly"
        intervalCount: 6
      }
    }
  ) {
    product {
      name
      id
      price
    }
    quantity
    trial
    grandTotal
  }
}
```

**Reply by tarikrital** [reply]
Now it gives the same error as I had before.

![image](https://github.com/user-attachments/assets/37e2fef7-e84d-44ac-9ee1-db1385bc2b2b)


Should I switch to `swell-node`? do you have an example using swell-node to achieve the same goal?

The product on the error has no relation with the current subscription of this operation, not on the plan nor on the invoice items!

**Reply by logeshswell** [reply]
Do you mean that this error is thrown only if the `quantity` is passed?

**Reply by logeshswell** [reply]
You need not switch to swell-node to update the quantity of subscription product. It should be done via graphQL or swell-js as well.

**Reply by tarikrital** [reply, marked as answer]
By reviewing the error message properly I found the issue with option, after setting it to null it works.
